### PR TITLE
Implement conditionalProgressRequester that allows requesting watch progress notification if watch cache is not fresh

### PR DESCRIPTION
**Pull request by serathius**
/kind feature
```release-note
NONE
```

Spin off https://github.com/kubernetes/kubernetes/pull/118508

/assign @wojtek-t

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress_test.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+	var requestsSent int32
+
+	logger.Info("No progress requests if no-one is waiting")
+	clock.Step(progressRequestPeriod * 2)
+
+	if err := wait.Poll(time.Millisecond, time.Second, func() (done bool, err error) {
+		requestsSent = pr.progressRequestsSentCount.Load()
+		return requestsSent == wantRequestsSent, nil
+	}); err != nil {
+		if requestsSent != wantRequestsSent {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress_test.go:51]: err != nil (with your function always returning nil) may only mean timeout - which means these are not equal.

So I would just change this if here to:
```
t.Errorf("Failed to wait for %d progress requests, got %d", ...)
```

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress_test.go:51]: Same in couple places below.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress_test.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+			t.Errorf("Number of progress requests didn't match, want %d, got %d", wantRequestsSent, requestsSent)
+		}
+	}
+
+	logger.Info("No progress after stopping")
+	close(stopCh)
+	time.Sleep(time.Millisecond)
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress_test.go:86]: remove

File: staging/src/k8s.io/apiserver/pkg/storage/interfaces.go [hunk 3]
```diff
@@ -236,6 +236,15 @@ type Interface interface {
 
 	// Count returns number of different entries under the key (generally being path prefix).
 	Count(key string) (int64, error)
+
+	// RequestBookmark requests etcd to send a bookmark to be sent on selected watches.
+	// In response etcd should send a bookmark informing the current progress on watch.
+	// Only watches with matching context grpc metadata (https://github.com/kubernetes/kubernetes/blob/9325a57125e8502941d1b0c7379c4bb80a678d5c/vendor/go.etcd.io/etcd/client/v3/watch.go#L1037-L1042)
+	// will be notified. Useful for resources with unfrequented changes with long periods of no events received.
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/interfaces.go:243]: Please update the comment to describe the semantic (the question I was asking - what if watch is lagging, etc.)

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_test.go [hunk 4]
```diff
@@ -112,19 +116,41 @@ func newTestWatchCache(capacity int, indexers *cache.Indexers) *testWatchCache {
 	}
 	versioner := storage.APIObjectVersioner{}
 	mockHandler := func(*watchCacheEvent) {}
-	wc := newWatchCache(keyFunc, mockHandler, getAttrsFunc, versioner, indexers, testingclock.NewFakeClock(time.Now()), schema.GroupResource{Resource: "pods"})
+	wc := &testWatchCache{}
+	wc.bookmarkRevision = make(chan int64, 1)
+	wc.stopCh = make(chan struct{})
+	pr := newConditionalProgressRequester(wc.RequestBookmark, clock.RealClock{})
```
[the remark this answers is not in the document]
Comment (serathius) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_test.go:122]: Ups, I didn't noticed that there is easily accessible fake clock. The test doesn't really use the clock, but let's have it consistent.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress.go [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+	cond    *sync.Cond
+	waiting int
+}
+
+type ProgressRequester func(ctx context.Context) error
+
+func (pr *conditionalProgressRequester) Run(stopCh <-chan struct{}) {
+	ctx := wait.ContextForChannel(stopCh)
+	go func() {
+		<-stopCh
```
[the remark this answers is not in the document]
Comment (serathius) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress.go:61]: Done

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress.go [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+		pr.mux.Lock()
+		defer pr.mux.Unlock()
+		pr.waiting = -1000000
```
[the remark this answers is not in the document]
Comment (serathius) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress.go:64]: Done.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress.go [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+	defer ticker.Stop()
+	for {
+		pr.mux.RLock()
+		for pr.waiting == 0 {
+			pr.cond.Wait()
+		}
+		pr.mux.RUnlock()
+		select {
+		case <-ticker.C():
+			pr.mux.RLock()
```
[the remark this answers is not in the document]
Comment (serathius) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress.go:77]: Done

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress_test.go (hunk 1 above)
Comment (serathius) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress_test.go:51]: Done

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress_test.go (hunk 2 above)
Comment (serathius) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress_test.go:86]: Done

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress_test.go [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+	go pr.Run(stopCh)
+	var wantRequestsSent int32
+	var requestsSent int32
+
+	logger.Info("No progress requests if no-one is waiting")
+	clock.Step(progressRequestPeriod * 2)
+
+	if err := wait.Poll(time.Millisecond, time.Second, func() (done bool, err error) {
+		requestsSent = pr.progressRequestsSentCount.Load()
+		return requestsSent == wantRequestsSent, nil
```
[the remark this answers is not in the document]
Comment (serathius) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress_test.go:49]: Added a wait of 20 milliseconds.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress.go [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+		pr.cond.Signal()
+	}()
+	ticker := pr.clock.NewTicker(progressRequestPeriod)
+	defer ticker.Stop()
+	for {
+		pr.mux.RLock()
+		for pr.waiting == 0 && !pr.stopped {
+			pr.cond.Wait()
+		}
+		pr.mux.RUnlock()
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress.go:77]: I would suggest explicitly adding:
```
if pr.stopped {
  return
}
``` 
here

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress_test.go [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+	*conditionalProgressRequester
+	progressRequestsSentCount atomic.Int32
+}
+
+func (pr *testConditionalProgressRequester) RequestProgress(ctx context.Context) error {
+	pr.progressRequestsSentCount.Add(1)
+	return nil
+}
+
+func PollConditionNoChange(interval, stable, timeout time.Duration, condition func() bool) {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress_test.go:116]: nit: make it private

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress_test.go [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+	"k8s.io/klog/v2"
+	"k8s.io/klog/v2/ktesting"
+	"k8s.io/utils/clock"
+	testingclock "k8s.io/utils/clock/testing"
+)
+
+var (
+	pollPeriod      = time.Millisecond
+	minimalNoChange = 20 * time.Millisecond
+	pollTimeout     = 100 * time.Millisecond
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress_test.go:36]: Let's make it bigger - 100ms is too low for heavily overloaded environments.

Given we're effectively never exercising pollTimeout anywhere (unless something actually goes wrong) let's make it 5s.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress_test.go [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+
+func (pr *testConditionalProgressRequester) RequestProgress(ctx context.Context) error {
+	pr.progressRequestsSentCount.Add(1)
+	return nil
+}
+
+func PollConditionNoChange(interval, stable, timeout time.Duration, condition func() bool) {
+	passCounter := 0
+	requiredNumberOfPasses := int(stable/interval) + 1
+	wait.Poll(interval, timeout, func() (done bool, err error) {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress_test.go:119]: this is silently eating error - change to:
```
return wait.Poll(...
```

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress_test.go [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+	pr := newTestConditionalProgressRequester(clock)
+	stopCh := make(chan struct{})
+	go pr.Run(stopCh)
+	var wantRequestsSent int32
+	var requestsSent int32
+
+	logger.Info("No progress requests if no-one is waiting")
+	clock.Step(progressRequestPeriod * 2)
+
+	PollConditionNoChange(pollPeriod, minimalNoChange, pollTimeout, func() bool {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress_test.go:53]:
```
if err := pollConditionNoChange(...
}); err != nil {
  t.Errorf("...")
}```

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress_test.go:53]: Same below

File: staging/src/k8s.io/apiserver/pkg/storage/interfaces.go (hunk 3 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/interfaces.go:243]: ping

**Review by wojtek-t — COMMENTED** [review]
Couple more minor comments, but this overall looks good.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress.go (hunk 9 above)
Comment (serathius) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress.go:77]: Done

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress_test.go (hunk 10 above)
Comment (serathius) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress_test.go:116]: Done

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress_test.go (hunk 11 above)
Comment (serathius) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress_test.go:36]: Done

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress_test.go (hunk 12 above)
Comment (serathius) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress_test.go:119]: Done

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress_test.go (hunk 13 above)
Comment (serathius) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress_test.go:53]: Done

File: staging/src/k8s.io/apiserver/pkg/storage/interfaces.go (hunk 3 above)
Comment (serathius) [reply staging/src/k8s.io/apiserver/pkg/storage/interfaces.go:243]: Sorry, was interrupted before sending the update.

**Review by wojtek-t — COMMENTED** [review]
Just one more small nit about the test itself - once fixed, this LGTM.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress_test.go [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+	var requestsSent int32
+
+	logger.Info("No progress requests if no-one is waiting")
+	clock.Step(progressRequestPeriod * 2)
+
+	if err := pollConditionNoChange(pollPeriod, minimalNoChange, pollTimeout, func() bool {
+		requestsSent = pr.progressRequestsSentCount.Load()
+		return requestsSent == wantRequestsSent
+	}); err != nil {
+		if requestsSent != wantRequestsSent {
```
[the remark this answers is not in the document]
Comment (serathius) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress_test.go:57]: Done

**Comment by serathius** [comment]
Done @wojtek-t

**Pull request merged**